Calculate Code Page ranges

m4rc1e opened the pull request:
Still mega wip. I'm basing my implementation on ufo2ft which in turn is based on FontForge.

m4rc1e commented:
This now produces the same result as the implementation in ufo2ft. I had to update `src/os2.rs` since the second codepage field was referencing the first.

I've added a simple test to check that the basic multilingual plane will enable all the specified codepages. I can make this more water tight if needed. However, due to this lib being in constant flux, it may be better to do this later.

@simoncozens I'm worrying about my overuse of clone and cloned when I'm using iterators and closures. Would be great to see what you would recommend.

m4rc1e commented:
Perhaps it may be better to move this to the os2 module since it computes the codepages using the cmap instead of ufo.info?

simoncozens commented:
That sounds like a nice idea: a method on an os2 that takes &mut self and a cmap table and fills in the relevant fields. Yes please!

moyogo commented:
This would also be useful as a single tool.

simoncozens commented:
Ping @cmyr

cmyr commented on crates/fonticulus/src/fontinfo.rs:
```diff
@@ -155,3 +156,135 @@ pub fn get_selection(info: &norad::FontInfo) -> u16 {
+    let unicodes_contains = | char | {
+        unicodes.contains(&(char as u32))
+    };
+    if unicodes_contains('Þ') && has_ascii {
+        code_page_ranges.insert(0); // Latin 1
+    }
+    if unicodes_contains('Ľ') && has_ascii {
+        code_page_ranges.insert(1); // Latin 2
+    }
+    if unicodes_contains('Б') {
+        code_page_ranges.insert(2); // Cyrillic
+        if unicodes_contains('Ѕ') && has_lineart {
+            code_page_ranges.insert(57); // IBM Cyrillic
+        }
+        if unicodes_contains('╜') && has_lineart {
+            code_page_ranges.insert(49); // MS-DOS Russian
+        }
+    }
```
now I'm just having fun, but I would be tempted to at least handle this simple case of "check char and addd to table" with a macro:

*some time later*:

Okay I'm going to not let myself rabbit-hole on this too hard, but as an illustration: with the following macro, I can replace this section with:

```rust
macro_rules! add_page_if_needed {
    // helper for if there's no condition and no block
    ($collection:ident, $ranges:ident, $chr:expr, $page:expr) => {
        add_page_if_needed!($collection, $chr, $page, true, {});
    };
    // helper for if there's a condition but no block
    ($collection:ident, $ranges:ident, $chr:expr, $page:expr, $constraint:expr) => {
        add_page_if_needed!($collection, $ranges, $chr, $page, $constraint, {});
    };
    // the verbose form
    ($collection:ident, $ranges:ident, $chr:expr, $page:expr, $constraint:expr, $block:block) => {
        if $collection.contains(&($chr as u32)) && $constraint {
            $ranges.insert($page);
            $block;
        }
    };
}
```

```suggestion
    add_page_if_needed!(unicodes, code_page_ranges,'Þ', 0, has_ascii);
    add_page_if_needed!(unicodes, code_page_ranges,'Ľ', 1, has_ascii);
    add_page_if_needed!(unicodes, code_page_ranges, 'Б', 2, true, {
        add_page_if_needed!(unicodes, code_page_ranges,'Ѕ', 57, has_lineart);
        add_page_if_needed!(unicodes, code_page_ranges, '╜', 49, has_lineart);
    });

```

In any case: this is provided more for illustrative purposes, and rust macros can be confusing, but I wanted to point this out as a way I would think about this problem. :)

cmyr reviewed (commented):
Okay, a bunch of notes inline.

As a general observation: `cargo fmt` and `cargo clippy` are very friendly, and I would encourage you to use them both; I'll look into turning them on in CI for this repo.

m4rc1e commented on crates/fonticulus/src/basictables.rs:
```diff
@@ -164,6 +165,15 @@ pub fn compile_os2(
         .open_type_os2_subscript_x_size
         .unwrap_or((upm * 0.65).round() as i32) as i16;
 
+    let unicodes:HashSet<u32> = mapping.keys().cloned().collect();
```
Thanks!

Which approach would you say is the best to adopt for the future? personally I find the turbofish style a little weird but for all I know this may be the most common.

cmyr replied:
I personally probably do the second form more often, but I wouldn't comment on the first form in code review 🤷

m4rc1e commented on crates/fonticulus/src/basictables.rs:
```diff
@@ -164,6 +165,15 @@ pub fn compile_os2(
         .unwrap_or((upm * 0.65).round() as i32) as i16;
 
+    let unicodes:HashSet<u32> = mapping.keys().cloned().collect();
+    let code_pages:Vec<u8> = info.open_type_os2_code_page_ranges.as_ref()
+        .unwrap_or(&calc_code_page_ranges(&unicodes))
+        .iter()
+        .cloned()
+        .collect();
```
Awesome thanks!

m4rc1e commented on crates/fonticulus/src/fontinfo.rs:
```diff
@@ -155,3 +156,135 @@ pub fn get_selection(info: &norad::FontInfo) -> u16 {
+/// implementation based on ufo2ft:
+/// https://github.com/googlefonts/ufo2ft/blob/main/lib/ufo2ft/util.py#l307
+pub fn calc_code_page_ranges(unicodes: &HashSet<u32>) -> Vec<u8> {
+    let mut code_page_ranges = HashSet::new();
```
> Is there a reason to prefer HashSet to Vec, here?

You're right, this could and probably should be a vec.

> would it make sense for us to generate the two separate vecs here, instead of doing it at the call-site? 

We could even go a step further and return a tuple containing two 32bit ints which is the desired end result.

cmyr replied:
oh totally, that's the sort of thing that rust excels at, and then we're doing way less work.

m4rc1e commented on crates/fonticulus/src/fontinfo.rs:
```diff
@@ -155,3 +156,135 @@ pub fn get_selection(info: &norad::FontInfo) -> u16 {
+    let mut code_page_ranges = HashSet::new();
+
+    let ascii_range: HashSet<u32> = (0x20..0x7E).collect();
+    let has_ascii = ascii_range.is_subset(&unicodes);
```
SGTM

m4rc1e commented on crates/fonticulus/src/basictables.rs:
```diff
@@ -164,6 +165,15 @@ pub fn compile_os2(
         .open_type_os2_subscript_x_size
         .unwrap_or((upm * 0.65).round() as i32) as i16;
 
+    let unicodes:HashSet<u32> = mapping.keys().cloned().collect();
```
I'm going to move this function so it becomes a method of the os2 table since it makes more sense for the code to live there. I may just end up making the method signature take the cmap table as input but I'll think about it.

m4rc1e commented:
Thank you very much for your review @cmyr! I hope it wasn't too painful. I've only been studying the existing code and the [Rust book](https://doc.rust-lang.org/book/). Are there other resources that you would recommend? I mainly come from a python background but have knowledge of C/C++. I find myself fighting the compiler a lot when it comes to membership and lifetimes so any links that helped you grok these topics quicker would be much appreciated.

cmyr commented:
I wish I had a good answer, but I largely don't. When I was becoming familiar with the language I spent a lot of time hanging out in a `#rust-beginners` irc channel; it doesn't exist anymore, but I believe there's a pretty active discord community? Another very useful resource is the [this week in rust](https://this-week-in-rust.org) newsletter; skimming that and reading things that interest you will be to just be ambiently aware of how other folks are writing the language.

Other than that, clippy and rustfmt really help avoid a lot of common problems; it really lets people avoid the bulk of argument over formatting and idiom, and clippy will often point out more idiomatic alternatives (like suggesting `my_vec.clear()` instead of `my_vec.truncate(0)`, as an example).

For the membership and lifetimes stuff: generally as folks write more rust they start to develop an intuition for a lot of this stuff. One thing that can be useful is to really try to imagine the stack, and imagine that things owned 'higher' in the call stack can be passed 'down' the stack, but things owned 'lower' can't be passed up, because when the function returns that memory is freed.

m4rc1e commented:
@cmyr I've implemented your feedback and moved most of the functionality to the os2 module. We know have two methods which can generate the codepage bits from a list of from a ufo's codepage attribute. Feel free to take another look. It isn't urgent. Once you're happy, I'll tackle the unicode ranges.

cmyr commented on src/os2.rs:
```diff
@@ -297,3 +299,139 @@ deserialize_visitor!(
+    }
+    /// implementation based on ufo2ft:
+    /// https://github.com/googlefonts/ufo2ft/blob/main/lib/ufo2ft/util.py#l307
+    pub fn calc_code_page_ranges(&mut self, mapping: &BTreeMap<u32, u16>) {
```
So now that I have a clearer idea of what we're doing, it feels like what we really want here is a function that creates two 32-bit integers in place. This is the sort of low-level thing that can often be annoying to do, but with Rust we could make this quite easy. What I would personally do is make a helper type that looks like a collection but is just doing bit arithmetic under the hood, something like:


```rust
#[derive(Default)]
struct CodePageRanges {
    one: u32,
    two: u32,
}

impl CodePageRanges {
    fn set(&mut self, idx: usize) {
        assert!(idx < 64);
        if idx >= 32 {
            self.two |= 1 << (idx - 32);
        } else {
            self.one |= 1 << idx;
        }
    }
    
    fn into_inner(self) -> (u32, u32) {
        (self.one, self.two)
    }
}
```

And then you can just call this directly, everwhere that you are using `code_page_ranges`:

```rust
let mut code_page_ranges = CodePageRanges::default();

if unicodes_contains('Þ') && has_ascii {
    code_page_ranges.set(0); // Latin 1
}
```

This then does everything at once; we're building up our final representation directly, as we add each new value.

m4rc1e replied:
Nice! this is much cleaner.

cmyr reviewed (commented):
A few more comments that you're totally free to ignore.

I  think that the code that you're writing works, and I think that's the most important thing; a lot of these other optimizations can come along later. That said, I have a sneaking suspicion that *somebody* is going to be using this library ten years in the future, so I don't think any effort spent on code quality etc is wasted time?

m4rc1e commented on crates/fonticulus/src/basictables.rs:
```diff
@@ -223,18 +224,24 @@ pub fn compile_os2(
         usUpperOpticalPointSize: None,
         fsSelection: get_selection(info),
+    };
+    if info.open_type_os2_code_page_ranges.is_none() {
+        table.calc_code_page_ranges(&mapping);
+    } else {
+        table.int_list_to_code_page_ranges(info.open_type_os2_code_page_ranges.as_ref().unwrap());
```
Thanks for your very detailed response.

I decided to convert the function into a method since I was planning on including a bin script which would recalculate the codepages for an existing font. I felt this was cleaner than the previous function implementation because I'd be duplicating a lot of the setup logic in the script. On the other hand, keeping it as a function means it isn't tied to a particular struct so we'll get improved reusability. I'm happy to revert back to a function based approach if Rune Bender etc will use these functions on sources.

cmyr replied:
I'm happy to defer to your judgement on this, and we can always make changes down the road as our needs reveal themselves.

m4rc1e commented:
@simoncozens feel free to merge this. I will implement the unicode range bits next so I may end up refactoring this part. I'm going to keep Colin's comments visible since they may help me later.

simoncozens commented:
Thanks very much for this - looks great!
